man/man-src/man-3.16.3/man-pages/man2/dprio.2
.\" Copyright (C) 2003 Free Software Foundation, Inc.
.\" This file is distributed according to the GNU General Public License.
.\" See the file COPYING in the top level source directory for details.
.\"
.\" Written by Sergey Oboguev.
.TH DPRIO 2 2014-09-05 "Linux" "Linux Programmer's Manual"
.SH NAME
dprio \- Deferred Set Task Priority
.SH "DESCRIPTION"

.BR dprio
is a facility to reduce the computational
cost of frequently-repeated
.BR sched_setattr (2)
system calls, intended for use by applications
that frequently execute this call.

Applications relying on fine-grain parallelism
may sometimes need to change their threads
priority at a very high rate, hundreds or even
thousands of times per typical scheduling
timeslice. These are typically applications that
have to execute short or very short lock-holding
critical or otherwise time-urgent sections of
code at a very high frequency and need to protect
these sections with "set priority" system calls,
one "set priority" call to elevate current thread
priority before entering the critical or
time-urgent section, followed by another call to
downgrade thread priority at the completion of
the section.

Due to the high frequency of entering and leaving
critical or time-urgent sections, the cost of
these "set priority" system calls may raise to a
noticeable part of an application's overall
expended CPU time.

.BR dprio
allows to largely eliminate the overhead of
these system calls.

Instead of executing a system call to elevate its
thread priority, an application simply writes its
desired priority level to a designated memory
area in the userspace. When the kernel attempts
to preempt the thread, it first checks the
content of this area, and if the application's
posted a request to change its priority
into the designated memory area, the
kernel will execute this request and alter the
priority of the thread being preempted before
performing a rescheduling, and then make
scheduling decisions based on the new thread
priority level thus implementing the priority
protection of the critical or time-urgent section
desired by the application.

In a predominant number of cases however, an
application will complete the critical section
before the end of the current timeslice and
cancel or alter the request held in the userspace
area. Thus a vast majority of an application's
change priority requests will be handled and
mutually cancelled or coalesced within the
userspace, at a very low overhead and without
incurring the cost of a system call, while
maintaining safe preemption control. The cost of
an actual kernel-level "set priority" operation
is incurred only if an application is actually
being preempted while inside the critical
section, i.e. typically at most once per
scheduling timeslice instead of hundreds or
thousands "set priority" system calls in the same
timeslice.

Application developers normally would not
interface to 
.BR dprio
directly and would rather use
a user-level library simplifying the use of the
.BR dprio
facility and shielding an application
developer from having to bother about low-level
details and reimplementing the boilerplate code
likely to be common and repetitive for all
applications using the deferred set priority
mechanism.

Below description of
.BR dprio
low-level interface is
thus of interest mostly to library and kernel
developers.

.BR dprio
facility consists of two parts.

One is the interface exposed via
.BR prctl (2)
with the option
.BR PR_SET_DEFERRED_SETPRIO.
It is used to set up the communication protocol
between the kernel and an area in an
application's memory space, as designated by the
application.
It is also used to terminate the protocol.

The other part is
.BR dprio
userspace <-> kernel
protocol itself, as described further below.


.SH DPRIO PRCTL INTERFACE

.BR prctl (2)
called with option
.BR PR_SET_DEFERRED_SETPRIO
has two forms: one to
set up the use of deferred set priority facility
for the current thread, another to terminate the
use of 
.BR dprio
for the thread.

The system call to set up the use of 
.BR dprio
for the thread takes the form

    prctl(PR_SET_DEFERRED_SETPRIO,
          dprio_ku_area_pp,
          sched_attrs_pp,
          sched_attrs_count,
          0)

The effect of this call is limited to the current
thread only.

\fIsched_attrs_pp\fP is a pointer to an array of
pointers to struct sched_attr (see
.BR sched_attr (2)). 
The array size is specified by
parameter \fIsched_attrs_count\fP. The array describes
a set of priorities (scheduling attributes) the
thread intends to subsequently request via
.BR dprio
mechanism and must contain at least one entry per
each scheduling policy (
.BR SCHED_NORMAL,
.BR SCHED_RR
etc.) that the application intends to
subsequently request via
.BR dprio.
Application can
specify multiple entries per scheduling policy in
the array, but only the entry with the highest
("best") priority for the given scheduling policy
really matters. For each of the scheduling
policies listed in the array,
.BR prctl(PR_SET_DEFERRED_SETPRIO)
will determine the
highest level of priority listed and verify
whether the calling thread is currently
authorized to use this level of priority. If the
thread is not authorized to use this priority,
.BR prctl
call will return error status. If yes,
.BR prctl(PR_SET_DEFERRED_SETPRIO)
will store inside
the kernel a pre-authorization for the thread to
subsequently elevate to this level of priority
via
.BR dprio.
.BR dprio
will also allow the thread to
elevate to recorded priority level and also lower
levels of priority within the same scheduling
policy.

The following scheduling policies can be listed
via \fIsched_attrs_pp\fP and subsequently used via
.BR dprio:
.BR SCHED_NORMAL,
.BR SCHED_IDLE,
.BR SCHED_BATCH,
.BR SCHED_RR
and
.BR SCHED_FIFO.
In most use cases an
application would use
.BR SCHED_FIFO 
and/or
.BR SCHED_RR,
although there are marginal use cases for other
listed classes as well.
.BR SCHED_DEADLINE
is not a
meaningful policy for use cases that
.BR dprio
is
intended for, and cannot be used with
.BR dprio.
An attempt to specify
.BR SCHED_DEADLINE
in
\fIsched_attrs_pp\fP will result in
.BR prctl(PR_SET_DEFERRED_SETPRIO)
returning an error.

More specifically, stored pre-authorization
consists of:

(a) For each scheduling policy listed in
\fIsched_attrs_pp\fP, the highest priority level
requested for this policy in \fIsched_attrs_pp\fP.

(b) A record of whether the caller had
.BR CAP_SYS_NICE
capability (either explicitly
assigned as a capability or caller having
effective id of root) at the time of the call.

If at any time after executing
.BR prctl(PR_SET_DEFERRED_SETPRIO)
thread priority
limits are subsequently constrained with
.BR prlimit/setrlimit(RLIMIT_RTPRIO)
or
.BR prlimit/setrlimit(RLIMIT_NICE),
new constraint
will affect the stored pre-authorization.
.BR dprio
will not allow the thread to elevate its priority
above the new limits set by
.BR prlimit/setrlimit(RLIMIT_RTPRIO)
and
.BR prlimit/setrlimit(RLIMIT_NICE)
regardless of the
pre-authorization created at the time of
.BR prctl(PR_SET_DEFERRED_SETPRIO),
unless the pre-authorization record also indicates that the
thread had
.BR CAP_SYS_NICE
capability at the time of
calling
.BR PR_SET_DEFERRED_SETPRIO.
This restriction
is intended to let external process management
applications to clamp down the application's
priority.

Thus calling
.BR PR_SET_DEFERRED_SETPRIO
creates a
limited pre-authorization context separate from
the thread's current security context and holding
a record of
.BR CAP_SYS_NICE
setting for the thread
of at the time of
.BR PR_SET_DEFERRED_SETPRIO
call.
If the thread decides to subsequently downgrade
its current security context and does not want
the code subsequent to this point to be able to
make use of
.BR CAP_SYS_NICE
recorded in the
.BR dprio
context, it is responsible for shutting down the
registered
.BR dprio
as well.

\fIdprio_ku_area_pp\fP is a pointer to u64 variable in
the userspace, let us name the latter
\fIdprio_ku_area_p\fP. This latter variable must be
aligned on u64 natural boundary, i.e. on the
8-byte boundary. When the application wants to
signal to the kernel its desire to change the
thread's priority, the application fills in the
desired priority settings and control data into
struct dprio_ku_area described below and stores
the address of the prepared struct dprio_ku_area
into \fIdprio_ku_area_p\fP. Even though \fIdprio_ku_area_p\fP
actually holds a pointer, it is declared as u64,
so the interface will be uniform regardless of
the system's bitness, and thus 32-bit
applications could interoperate with 64-bit
kernel. On 32-bit systems, the upper part of
\fIdprio_ku_area_p\fP is left zero.

.BR prctl(PR_SET_DEFERRED_SETPRIO)
stores the value
of \fIdprio_ku_area_pp\fP inside the kernel. When the
kernel subsequently attempts to preempt the
thread, it checks the stored value of
\fIdprio_ku_area_pp\fP, and if it not NULL, tries to
fetch from the userspace a u64 value pointed to
by \fIdprio_ku_area_pp\fP, i.e. \fIdprio_ku_area_p\fP. If the
fetch attempt fails (either because
\fIdprio_ku_area_pp\fP points to an invalid memory or
page has been paged out), the kernel ignores the
.BR dprio
setting and proceeds as if the thread was
not set up for
.BR dprio.
If the fetched value of
\fIdprio_ku_area_p\fP is NULL, there is no
.BR dprio
request from the application and the kernel
proceeds with rescheduling. If the fetched value
of \fIdprio_ku_area_p\fP is not NULL, the kernel
fetches struct dprio_ku_area pointed by
\fIdprio_ku_area_p\fP. This area has the following
structure defined by <linux/dprio_api.h>:
.nf

    #include <linux/dprio_api.h>

    struct dprio_ku_area {
    	    __dprio_volatile __u32 resp;      /* DPRIO_RESP_xxx */
	    __dprio_volatile __u32 error;     /* one of errno values */
	    __dprio_volatile struct sched_attr sched_attr;
    };

.fi
The value of __dprio_volatile may be controlled by the
application including <linux/dprio_api.h> depending on
the method of access to the structure utilized by the application.
Application may define __dprio_volatile to be
.BR volatile
or it may leave it empty and utilize compiler barriers instead.

If the new priority setting requested by the
application in sched_attr field fits within the
pre-authorization stored for the thread during
.BR prctl(PR_SET_DEFERRED_SETPRIO),
the kernel will
alter the thread's priority and scheduling
attributes to the new value requested in
sched_attr.

The kernel will try to store the success/error
status of the operation in resp and error fields
and attempt to reset the value of \fIdprio_ku_area_p\fP
to NULL. The exact kernel <-> userspace
.BR dprio
communication protocol is described in a separate
section below.

The system call to terminate the use of
.BR dprio
for the thread takes the form

    prctl(PR_SET_DEFERRED_SETPRIO, 0, 0, 0, 0)

After executing this call, the kernel will clear
any previously stored value of \fIdprio_ku_area_pp\fP
for the thread and will no longer attempt to
check for
.BR dprio
requests for the thread.

It is crucial that the user of
.BR dprio
facility remembers to detach from
.BR dprio
after stopping to
use it and before relinquishing the ownership of
memory areas pointed by \fIdprio_ku_area_pp\fP and the
pointer stored in \fIdprio_ku_area_pp\fP, as the kernel
will continue reading from and writing to these
areas of memory until
.BR dprio
use is terminated
either via
.BR prctl (2)
or thread termination.
Therefore relinquishing the ownership of these
memory areas prior to terminating their
designation for
.BR dprio
will likely result in (1)
task memory corruption and (2) the possibility of
unintended task priority changes. The use of
.BR dprio
for the thread terminates automatically if
the thread terminates.

Right before detaching the thread from the
previously designated \fIdprio_ku_area_pp\fP the kernel
will make one last attempt to check for a pending
.BR dprio
request associated with the previous value
of \fIdprio_ku_area_pp\fP and process any currently
pending
.BR dprio
request pointed to by this value.

Likewise, if the application calls

    prctl(PR_SET_DEFERRED_SETPRIO, dprio_ku_area_1_pp, ...)
    . . . . .
    prctl(PR_SET_DEFERRED_SETPRIO, dprio_ku_area_2_pp, ...)

then right before switching over to
\fIdprio_ku_area_2_pp\fP, the kernel will process a
pending
.BR dprio
request pointed by \fIdprio_ku_area_1_pp\fP.

.BR dprio
requests are handled both on voluntary and
involuntary preemption of the thread, i.e. if the
thread has a
.BR dprio
request posted, the request
will be handled when the thread is preempted by a
higher-priority task or on a round-robin basis,
but also when the thread enters wait state, e.g.
by calling sleep(3) or trying to read from a
socket with no data available or doing epoll etc.

.BR PR_SET_DEFERRED_SETPRIO
setting is per-thread and
is not inherited by child threads or processes
created via
.BR clone (2)
or
.BR fork (2).
Only the effects
of priority change requests issued via
.BR dprio
prior to
.BR clone (2)
or
.BR fork (2)
have an effect on a
child task or thread (as long as this is not
restricted by
.BR SCHED_RESET_ON_FORK
or
.BR SCHED_FLAG_RESET_ON_FORK),
but a child task or
thread does not inherit the parent's
.BR PR_SET_DEFERRED_SETPRIO
setting.

.BR PR_SET_DEFERRED_SETPRIO
setting is reset by
.BR execve (2).
Only the effects of priority change
requests issued via
.BR dprio
prior to
.BR execve (2)
have an effect on the loaded executable image, but a
new image executed by the task does not inherit
.BR PR_SET_DEFERRED_SETPRIO
setting.

When executing
.BR clone (2),
.BR fork (2)
or
.BR execve (2)
the kernel will check for a pending
.BR dprio
request and
try to process it before executing the main
syscall body, so the effects of priority
adjustment request posted via
.BR dprio
will be
integrated in the outcome of the mentioned
syscalls.

When executing a
.BR dprio
request, the kernel will
try to merge
.BR SCHED_FLAG_RESET_ON_FORK
into the
current task state, as follows.

If the task does not have the
.BR SCHED_RESET_ON_FORK
flag set, and the request does have the
.BR SCHED_FLAG_RESET_ON_FORK
flag set in sched_attr
structure, the kernel will set the
.BR SCHED_RESET_ON_FORK
flag for the task.

If the task already has the
.BR SCHED_RESET_ON_FORK
flag set, but the request does not have the
.BR SCHED_FLAG_RESET_ON_FORK
flag set in sched_attr
structure, the task will retain the flag.

The resultant task's "reset on fork" flag state
is the logical "OR" of the preceding task's flag
state and the
.BR SCHED_FLAG_RESET_ON_FORK
flag state
in the request. If the flag in the request is not
set, the kernel will not attempt to reset the
task's flag.

.SH DPRIO AUTHORIZATION

.BR dprio
is an optional facility and may be present
in the system or not depending on the system build
options.

System administrator may also designate
.BR dprio
as
requiring CAP_DPRIO capability to use it.
Privileged or non-privileged status of
.BR dprio
can
be specified at system build time and then
dynamically altered by system administrator via
\fI/proc/sys/kernel/dprio_privileged\fP or
\fIsysctl kernel.dprio_privileged\fP.

.SH DPRIO PRCTL ERRORS

.BR prctl(PR_SET_DEFERRED_SETPRIO)
may return the following errors:

.B
EINVAL
This system does not support
.BR dprio,
or one
of the entries in \fIsched_attrs_pp\fP array lists an
unsupported scheduling policy, or a reserved
.BR prctl (2)
argument is not specified as 0, or
\fIdprio_ku_area_pp\fP does not point location that is
readable, writable and aligned on 8-byte
boundary.

.B EPERM
Thread is not authorized to use a priority
setting it has listed in \fIsched_attrs_pp\fP array.

.B EFAULT
One of the parameters points outside of
accessible address space.

.B E2BIG
One of struct sched_setattr structures
pointed from \fIsched_attrs_pp\fP array is malformatted.

.B ENOMEM
The system is out of memory and cannot
allocate a pre-authorization context.

.SH DPRIO USERSPACE <-> KERNEL PROTOCOL

Userspace <-> kernel can be used after
.BR dprio
userspace <-> kernel is established with
.BR prctl(PR_SET_DEFERRED_SETPRIO).

.BR dprio
protocol is defined as
follows.

To post a deferred set priority request,
userspace performs:

    Select and fill-in dprio_ku_area:

        Set \fIresp\fP = DPRIO_RESP_NONE.
        Set \fIsched_attr\fP.

    Set \fIdprio_ku_area_p\fP to point struct dprio_ku_area.

Kernel:

.TP
1)
On task preemption attempt or at another
processing point, such as fork or exec, read
\fIdprio_ku_area_p\fP. If \fIdprio_ku_area_p\fP is not
readable
(inaccessible incl. page swapped out),
quit.

Note: will reattempt again on next
preemption cycle.

.TP
2)
If read-in value of \fIdprio_ku_area_p\fP is 0,
do nothing and quit.

.TP
3)
Set \fIresp\fP = DPRIO_RESP_UNKNOWN.
.IP
If cannot (e.g. \fIresp\fP inaccessible), quit.

.TP
4)
Set \fIdprio_ku_area_p\fP = NULL.
.IP
If cannot (e.g. inaccessible), quit.
.IP
Note that in this case request handling
will be reattempted on next
thread preemption cycle. Thus \fIresp\fP value
of DPRIO_RESP_UNKNOWN may
be transient and overwritten with
DPRIO_RESP_OK or DPRIO_RESP_ERROR
if \fIdprio_ku_area_p\fP is not reset to 0 by
the kernel (or to 0 or to
the address of another dprio_ku_area by
the userspace).

.TP
5)
Read \fIsched_attr\fP.
.IP
If cannot (e.g. inaccessible), quit.

.TP
6)
Try to change task scheduling attributes
in accordance with read-in
value of \fIsched_attr\fP.

.TP
7)
If successful, set \fIresp\fP = DPRIO_RESP_OK
and quit.

.TP
8)
If unsuccessful, set \fIerror\fP = appopriate
errno-style value.
.IP
If cannot (e.g. \fIerror\fP inaccessible), quit.
.IP
Set \fIresp\fP = DPRIO_RESP_ERROR.
.IP
If cannot (e.g. \fIresp\fP inaccessible), quit.

.P
Explanation of possible \fIresp\fP codes:

.TP
DPRIO_RESP_NONE

src/linux-3.16.3/kernel/sched/dprio.c
/*
 * kernel/sched/dprio.c
 *
 * Deferred set priority.
 *
 * Started by (C) 2014 Sergey Oboguev <oboguev@yahoo.com>
 *
 * This code is licenced under the GPL version 2 or later.
 * For details see linux-kernel-base/COPYING.
 */

#include <linux/types.h>
#include <linux/unistd.h>
#include <linux/stddef.h>
#include <linux/errno.h>
#include <linux/kernel.h>
#include <linux/sched.h>
#include <linux/dprio.h>
#include <linux/dprio_api.h>
#include <linux/slab.h>
#include <linux/compiler.h>
#include <linux/uaccess.h>
#include <linux/capability.h>
#include <linux/prctl.h>
#include <linux/init.h>

unsigned int dprio_privileged = DPRIO_PRIVILEGED_INITIAL_VALUE;

/*
 * Returns 0 on success.
 */
static inline int __copyin(void *dst, const void __user *src,
			   unsigned size, bool atomic)
{
	int ret;

	/* Use barrier() to sequence userspace-kernel dprio protocol */
	barrier();
	if (atomic) {
		pagefault_disable();
		ret = __copy_from_user_inatomic(dst, src, size);
		pagefault_enable();
	} else {
		ret = copy_from_user(dst, src, size);
	}
	barrier();

	return ret;
}

/*
 * Returns 0 on success.
 */
static inline int __copyout(void __user *dst, const void *src,
			    unsigned size, bool atomic)
{
	int ret;

	/* Use barrier() to sequence userspace-kernel dprio protocol */
	barrier();
	if (atomic) {
		pagefault_disable();
		ret = __copy_to_user_inatomic(dst, src, size);
		pagefault_enable();
	} else {
		ret = copy_to_user(dst, src, size);
	}
	barrier();

	return ret;
}

#define __copyin_var(x, uptr, atomic)	\
	__copyin(&(x), (uptr), sizeof(x), (atomic))

#define __copyout_var(x, uptr, atomic)	\
	__copyout((uptr), &(x), sizeof(x), (atomic))


/*
 * Mimics sched_copy_attr()
 */
#define CHUNK_SIZE 32u
static int dprio_copyin_sched_attr(struct sched_attr __user *uattr,
				   struct sched_attr *attr,
				   bool atomic)
{
	u32 size;

	if (!access_ok(VERIFY_READ, uattr, SCHED_ATTR_SIZE_VER0))
		return -EFAULT;

	/*
	 * zero the full structure, so that a short copy will be nice.
	 */
	memset(attr, 0, sizeof(*attr));

	if (__copyin_var(size, &uattr->size, atomic))
		return -EFAULT;

	if (size > PAGE_SIZE)	/* silly large */
		return -E2BIG;

	if (!size)		/* abi compat */
		size = SCHED_ATTR_SIZE_VER0;

	if (size < SCHED_ATTR_SIZE_VER0)
		return -E2BIG;

	/*
	 * If we're handed a bigger struct than we know of,
	 * ensure all the unknown bits are 0 - i.e. new
	 * user-space does not rely on any kernel feature
	 * extensions we dont know about yet.
	 */
	if (size > sizeof(*attr)) {
		unsigned char __user *addr;
		unsigned char __user *end;
		unsigned char val[CHUNK_SIZE];
		unsigned k, chunk_size;

		addr = (char __user *)uattr + sizeof(*attr);
		end  = (char __user *)uattr + size;

		for (; addr < end; addr += chunk_size) {
			chunk_size = min((unsigned) (end - addr), CHUNK_SIZE);
			if (__copyin(val, addr, chunk_size, atomic))
				return -EFAULT;
			for (k = 0;  k < chunk_size; k++) {
				if (val[k])
					return -E2BIG;
			}
		}
		size = sizeof(*attr);
	}

	if (__copyin(attr, uattr, size, atomic))
		return -EFAULT;

	attr->size = size;

	/*
	 * XXX: do we want to be lenient like existing syscalls; or do we want
	 * to be strict and return an error on out-of-bounds values?
	 * See also other uses of clamp(..., MIN_NICE, MAX_NICE) below.
	 */
	attr->sched_nice = clamp(attr->sched_nice, MIN_NICE, MAX_NICE);

	return 0;
}


/*
 * Detach the task from userland deferred setprio request area and deallocate
 * all resources for the connection. Called from:
 *
 *   - prctl(PR_SET_DEFERRED_SETPRIO) with area argument passed as NULL
 *     to terminate previous connection
 *
 *   - prctl(PR_SET_DEFERRED_SETPRIO) with new non-NULL area argument
 *     setting new connection. Previous connection is terminated before
 *     establishing a new one
 *
 *   - when the task is terminated in do_exit()
 */
void dprio_detach(struct task_struct *tsk)
{
	preempt_disable();

	tsk->dprio_ku_area_pp = NULL;

	if (unlikely(tsk->dprio_info)) {
		kfree(tsk->dprio_info);
		tsk->dprio_info = NULL;
	}

	preempt_enable();
}

/*
 * Pre-process sched_attr just read from the userspace, whether during precheck
 * or during dprio request execution, to impose uniform interpretation of
 * structure format and values.
 */
static void uniform_attr(struct sched_attr *attr)
{
	/* accommodate legacy hack */
	if ((attr->sched_policy & SCHED_RESET_ON_FORK) &&
	    attr->sched_policy != -1) {
		attr->sched_flags |= SCHED_FLAG_RESET_ON_FORK;
		attr->sched_policy &= ~SCHED_RESET_ON_FORK;
	}

	if (attr->sched_policy == SCHED_IDLE)
		attr->sched_nice = MAX_NICE;
}

/*
 * Precheck whether current process is authorized to set its scheduling
 * properties to @uattr. If yes, make record in @info and return 0.
 * If not, return error.
 */
static int precheck(struct dprio_info *info, struct sched_attr __user *uattr)
{
	struct sched_attr attr;
	u32 policy;
	unsigned mask;
	int error;

	error = dprio_copyin_sched_attr(uattr, &attr, false);
	if (error)
		return error;

	uniform_attr(&attr);

	policy = attr.sched_policy;
	mask = 1 << policy;

	switch (policy) {
	case SCHED_NORMAL:
		attr.sched_nice = clamp(attr.sched_nice, MIN_NICE, MAX_NICE);
		if ((info->mask & mask) &&
		    attr.sched_nice >= info->normal_sched_nice)
			break;
		error = sched_setattr_precheck(current, &attr);
		if (error == 0) {
			info->normal_sched_nice = attr.sched_nice;
			info->mask |= mask;
		}
		break;

	case SCHED_BATCH:
		attr.sched_nice = clamp(attr.sched_nice, MIN_NICE, MAX_NICE);
		if ((info->mask & mask) &&
		    attr.sched_nice >= info->batch_sched_nice)
			break;
		error = sched_setattr_precheck(current, &attr);
		if (error == 0) {
			info->batch_sched_nice = attr.sched_nice;
			info->mask |= mask;
		}
		break;

	case SCHED_FIFO:
		if ((info->mask & mask) &&
		    attr.sched_priority <= info->fifo_sched_priority)
			break;
		error = sched_setattr_precheck(current, &attr);
		if (error == 0) {
			info->fifo_sched_priority = attr.sched_priority;
			info->mask |= mask;
		}
		break;

	case SCHED_RR:
		if ((info->mask & mask) &&
		    attr.sched_priority <= info->rr_sched_priority)
			break;
		error = sched_setattr_precheck(current, &attr);
		if (error == 0) {
			info->rr_sched_priority = attr.sched_priority;
			info->mask |= mask;
		}
		break;

	case SCHED_IDLE:
		if (info->mask & mask)
			break;
		error = sched_setattr_precheck(current, &attr);
		if (error == 0)
			info->mask |= mask;
		break;

	case SCHED_DEADLINE:
		/*
		 * DL is not a meaningful policy for deferred set
		 * priority
		 */
	default:
		error = -EINVAL;
		break;
	}

	return error;
}

/*
 * Implements prctl(PR_SET_DEFERRED_SETPRIO).
 *
 * To set PR_SET_DEFERRED_SETPRIO:
 *
 *     a2 = address of u64 variable in the userspace that holds the pointer
 *          to dprio_ku_area or NULL
 *
 *     a3 = address of userspace array of pointers to sched_attr entries
 *          to preapprove for subsequent pre-checked use by deferred set
 *          priority requests
 *
 *     a4 = count of entries in a3 or 0
 *
 *     a5 = 0
 *
 * To reset PR_SET_DEFERRED_SETPRIO:
 *
 *     a2 = 0
 *     a3 = 0
 *     a4 = 0
 *     a5 = 0
 *
 * Thus valid calls are:
 *
 *     struct sched_attr **sched_attrs_pp;
 *     prctl(PR_SET_DEFERRED_SETPRIO, dprio_ku_area_pp,
 *           sched_attrs_pp, nattrs, 0)
 *
 *     prctl(PR_SET_DEFERRED_SETPRIO, NULL, NULL, 0, 0)
 *
 */
long dprio_prctl(int option, unsigned long a2, unsigned long a3,
		 unsigned long a4, unsigned long a5)
{
	struct dprio_ku_area __user * __user *ku_area_pp;
	struct dprio_ku_area __user *ku_area_p;
	struct dprio_info *info = NULL;
	unsigned long ne, nentries;
	struct sched_attr __user * __user *uattr_pp;
	struct sched_attr __user *uattr_p;
	bool atomic = false;
	long error = 0;

	if (option != PR_SET_DEFERRED_SETPRIO)
		return -EINVAL;

	ku_area_pp = (struct dprio_ku_area __user * __user *) a2;

	/*
	* Handle reset operation for PR_SET_DEFERRED_SETPRIO
	 */
	if (ku_area_pp == NULL) {
		if (a3 | a4 | a5)
			return -EINVAL;
		dprio_handle_request();
		dprio_detach(current);
		return 0;
	}

	/*
	 * Handle set operation for PR_SET_DEFERRED_SETPRIO
	 */
	uattr_pp = (struct sched_attr __user * __user *) a3;
	nentries = a4;
	if (a5)
		return -EINVAL;

	/* sanity check to avoid long spinning in the kernel */
	if (nentries > 4096) {
		error = -EINVAL;
		goto out;
	}

	/* Check alignment */
	if ((unsigned long) ku_area_pp % sizeof(u64))
		return  -EINVAL;

	/* check *ku_area_pp is readable and writeable */
	if (__copyin_var(ku_area_p, ku_area_pp, atomic) ||
	    __copyout_var(ku_area_p, ku_area_pp, atomic))
		return  -EFAULT;

	error = dprio_check_permission();
	if (error)
		return error;

	info = kmalloc(sizeof(*info), GFP_KERNEL);
	if (info == NULL)
		return -ENOMEM;
	info->mask = 0;
	/*
	 * XXX:
	 *
	 * We may trigger a false recording of PF_SUPERPRIV here by requesting
	 * CAP_SYS_NICE capability we may not actually use later, however
	 * since we cannot modify current->flags during dprio_handle_request()
	 * when called from __schedule(), the alternatives would be either
	 * possibly missing the recording of PF_SUPERPRIV, or (better) splitting
	 * PF_SUPERPRIV from current->flags and moving it to a variable with
	 * atomic access protocol.
	 */
	info->capable_sys_nice = capable(CAP_SYS_NICE);

	/*
	 * We prevalidate maximum requested priority levels at the time of
	 * prctl set-up instead of validating priority change requests during
	 * their actual processing in __schedule and do_fork in order to:
	 *
	 *    - reduce latency during request processing in __schedule()
	 *
	 *    - avoid blocking in the secirity code when setprio processing
	 *      is performed in _schedule()
	 *
	 *    - avoid EINTR or ERESTARTSYS etc. that may be returned by
	 *      the security code during setprio request processing
	 */
	for (ne = 0;  ne < nentries;  ne++) {
		cond_resched();
		if (__copyin_var(uattr_p, uattr_pp + ne, atomic)) {
			error = -EFAULT;
			goto out;
		}
		error = precheck(info, uattr_p);
		if (error)
			goto out;
	}

	/*
	 * If there was a previous active dprio ku area, try to process
	 * any pending request in it and detach from it.
	 */
	dprio_handle_request();
	dprio_detach(current);

	preempt_disable();
	current->dprio_ku_area_pp = ku_area_pp;
	current->dprio_info = info;
	preempt_enable();

out:
	if (error && info)
		kfree(info);

	return error;
}

/*
 * Check if "deferred set priority" request from the userland is pending.
 * Returns @true if request has been detected, @false if not.
 *
 * If page pointed by dprio_ku_area_pp is not currently accessible (e.g. not
 * valid or paged out), return @false.
 */
bool dprio_check_for_request(struct task_struct *prev)
{
	struct dprio_ku_area __user *ku_area_p;
	bool atomic = true;

#ifdef CONFIG_DEBUG_DEFERRED_SETPRIO
	/*
	 * We are only called if prev->dprio_ku_area_pp != NULL,
	 * thus prev cannot be a kernel thread
	 */
	if (unlikely(prev->active_mm != prev->mm)) {
		WARN_ONCE(1, KERN_ERR "BUG: dprio: address space not mapped\n");
		return false;
	}
#endif /* CONFIG_DEBUG_DEFERRED_SETPRIO */

	if (__copyin_var(ku_area_p, prev->dprio_ku_area_pp, atomic))
		return false;

	return ku_area_p != NULL;
}

/*
 * Handle pending "deferred set priority" request from the userland.
 */
void dprio_handle_request(void)
{
	struct dprio_ku_area __user *ku;
	struct dprio_ku_area __user *ku_null;
	struct sched_attr attr;
	bool atomic;
	u32 resp, error;
	int ierror = 0;
	unsigned long rlim_rtprio;
	long rlim_nice;
	struct dprio_info *info;

	/* attached to ku area? */
	if (current->dprio_ku_area_pp == NULL)
		return;

	/* called from __schedule? */
	atomic = preempt_count() != 0;

	/* fetch ku request area address from the userspace */
	if (__copyin_var(ku, current->dprio_ku_area_pp, atomic))
		return;

	/* check if request is pending */
	if (unlikely(ku == NULL))
		return;

	/* remark to the userspace:
	   request processing has been started/attempted */
	resp = DPRIO_RESP_UNKNOWN;
	if (__copyout_var(resp, &ku->resp, atomic))
		return;

	/* reset pending request */
	ku_null = NULL;
	if (__copyout_var(ku_null, current->dprio_ku_area_pp, atomic))
		return;

	/* fetch request parameters from the userspace */
	if (dprio_copyin_sched_attr(&ku->sched_attr, &attr, atomic))
		return;

	/* impose uniform interpretation of sched_attr */
	uniform_attr(&attr);

	if (attr.sched_flags & ~SCHED_FLAG_RESET_ON_FORK) {
		ierror = -EINVAL;
		goto out;
	}

	/*
	 * check if request has been pre-authorized
	 */
	info = current->dprio_info;
	switch (attr.sched_policy) {
	case SCHED_NORMAL:
		if (!(info->mask & (1 << SCHED_NORMAL)) ||
		    attr.sched_nice < info->normal_sched_nice)
			ierror = -EPERM;
		/*
		 * check whether RLIMIT_NICE has been reduced
		 * by setrlimit or prlimit
		 */
		if (ierror == 0 && !info->capable_sys_nice) {
			rlim_nice = 20 - task_rlimit(current, RLIMIT_NICE);
			if (attr.sched_nice < rlim_nice)
				ierror = -EPERM;
		}
		break;

	case SCHED_BATCH:
		if (!(info->mask & (1 << SCHED_BATCH)) ||
		    attr.sched_nice < info->batch_sched_nice)
			ierror = -EPERM;
		/*
		 * check whether RLIMIT_NICE has been reduced
		 * by setrlimit or prlimit
		 */
		if (ierror == 0 && !info->capable_sys_nice) {
			rlim_nice = 20 - task_rlimit(current, RLIMIT_NICE);
			if (attr.sched_nice < rlim_nice)
				ierror = -EPERM;
		}
		break;

	case SCHED_FIFO:
		if (!(info->mask & (1 << SCHED_FIFO)) ||
		    attr.sched_priority > info->fifo_sched_priority)
			ierror = -EPERM;
		/*
		 * check whether RLIMIT_RTPRIO has been reduced
		 * by setrlimit or prlimit
		 */
		if (ierror == 0 && !info->capable_sys_nice) {
			rlim_rtprio = task_rlimit(current, RLIMIT_RTPRIO);
			if (rlim_rtprio == 0 || attr.sched_priority > rlim_rtprio)
				ierror = -EPERM;
		}
		break;

	case SCHED_RR:
		if (!(info->mask & (1 << SCHED_RR)) ||
		    attr.sched_priority > info->rr_sched_priority)
			ierror = -EPERM;
		/*
		 * check whether RLIMIT_RTPRIO has been reduced
		 * by setrlimit or prlimit
		 */
		if (ierror == 0 && !info->capable_sys_nice) {
			rlim_rtprio = task_rlimit(current, RLIMIT_RTPRIO);
			if (rlim_rtprio == 0 || attr.sched_priority > rlim_rtprio)
				ierror = -EPERM;
		}
		break;

	case SCHED_IDLE:
		if (!(info->mask & (1 << SCHED_IDLE)))
			ierror = -EPERM;
		break;

	default:
		ierror = -EINVAL;
		break;
	}

	/* execute the request */
	if (ierror == 0)
		ierror = sched_setattr_prechecked(current, &attr, true);

out:
	if (ierror) {
		error = (u32) -ierror;
		resp = DPRIO_RESP_ERROR;
		if (0 == __copyout_var(error, &ku->error, atomic))
			__copyout_var(resp, &ku->resp, atomic);
	} else {
		resp = DPRIO_RESP_OK;
		__copyout_var(resp, &ku->resp, atomic);
	}
}

/*
 * Verify if the current task is authorized to use prctl(PR_SET_DEFERRED_SETPRIO).
 */
int dprio_check_permission(void)
{
	if (dprio_privileged && !capable(CAP_DPRIO))
		return -EPERM;

	return 0;
}

